arches/app/models/querysets.py

```diff
@@ -0,0 +1,243 @@
+from django.db.models import OuterRef, Prefetch, QuerySet
+
+from arches.app.models.utils import find_root_node, generate_tile_annotations
+
+
+class TileQuerySet(QuerySet):
+    def with_node_values(
+        self, nodes, *, defer=None, only=None, lhs=None, outer_ref, depth=1
+    ):
+        """TileModel.as_nodegroup() is the better entrypoint, see docs there."""
+        from arches.app.models.models import TileModel
+
+        node_alias_annotations = generate_tile_annotations(
+            nodes,
+            defer=defer,
+            only=only,
+            model=self.model,
+            lhs=lhs,
+            outer_ref=outer_ref,
+        )
+
+        prefetches = []
+        if depth:
+            prefetches.append(
+                Prefetch(
+                    "children",
+                    queryset=TileModel.objects.with_node_values(
+                        nodes,
+                        defer=defer,
+                        only=only,
+                        depth=depth - 1,
+                        lhs="parenttile",
+                        outer_ref="tileid",
+                    ),
+                )
+            )
+
+        self._fetched_nodes = [n for n in nodes if n.alias in node_alias_annotations]
+        return (
+            self.filter(data__has_any_keys=[n.pk for n in self._fetched_nodes])
+            .prefetch_related(*prefetches)
+            .annotate(
+                **node_alias_annotations,
+            )
+            .order_by("sortorder")
+        )
+
+    def _prefetch_related_objects(self):
+        """Call datatype to_python() methods when materializing the QuerySet.
+        Discard annotations that do not pertain to this nodegroup.
+        """
+        from arches.app.datatypes.datatypes import DataTypeFactory
+
+        super()._prefetch_related_objects()
+
+        datatype_factory = DataTypeFactory()
+        NOT_PROVIDED = object()
+        for tile in self._result_cache:
+            for node in self._fetched_nodes:
+                if node.nodegroup_id == tile.nodegroup_id:
+                    tile._root_node = node
+                    tile_val = getattr(tile, node.alias, NOT_PROVIDED)
+                    if tile_val is not NOT_PROVIDED:
+                        datatype_instance = datatype_factory.get_instance(node.datatype)
+                        python_val = datatype_instance.to_python(tile_val)
+                        setattr(tile, node.alias, python_val)
```

Comment:
Er, probably just on the `__getattr__`, and treat anything accessed as potentially dirty, because you can directly mutate things without setting something on the python object directly (unless we consider that a feature, that such an update is blocked...)